livekit-agents/livekit/agents/utils/message_change.py

```diff
@@ -0,0 +1,177 @@
+from dataclasses import dataclass
+from typing import Callable, Generic, TypeVar
+
+T = TypeVar("T")
+
+
+@dataclass
+class MessageChange(Generic[T]):
+    """Represents changes needed to transform one list into another
+
+    The changes must be applied in order:
+    1. First apply all deletions
+    2. Then apply all insertions with their previous_item_id
+    """
+
+    to_delete: list[T]
+    """Items to delete from old list"""
+    to_add: list[tuple[T | None, T]]
+    """Items to add as (previous_item, new_item) pairs"""
+
+
+def compute_changes(
+    old_list: list[T], new_list: list[T], key_fnc: Callable[[T], str]
+) -> MessageChange[T]:
+    """Compute minimum changes needed to transform old list into new list"""
+    # Convert to lists of ids
+    old_ids = [key_fnc(msg) for msg in old_list]
+    new_ids = [key_fnc(msg) for msg in new_list]
+
+    # Create lookup maps
+    old_msgs = {key_fnc(msg): msg for msg in old_list}
+    new_msgs = {key_fnc(msg): msg for msg in new_list}
+
+    # Compute changes using ids
+    changes = _compute_list_changes(old_ids, new_ids)
+
+    # Convert back to items
+    return MessageChange(
+        to_delete=[old_msgs[id] for id in changes.to_delete],
+        to_add=[
+            (
+                None if prev is None else old_msgs.get(prev) or new_msgs[prev],
+                new_msgs[new],
+            )
+            for prev, new in changes.to_add
+        ],
+    )
+
+
+def _compute_list_changes(old_list: list[T], new_list: list[T]) -> MessageChange[T]:
+    """Compute minimum changes needed to transform old_list into new_list
+
+    Rules:
+    - Delete first, then insert
+    - Can't insert at start if list not empty (must delete all first)
```

Comment:
To add to this comment, I'm mostly mentioning it because you can also send "audio data" as part of the message (for user messages). So it requires more bandwidth

Reply:
It will delete all items only if we want to insert a new item to the first. It seems that the API can only insert "after an item" with the `previous_item_id` so if the session is not empty there is no way to insert to the beginning. Please let me know if there is a way to do that.

Reply:
Ah this makes sense, kind of weird their API doesn't support to prepend an item 😅